### PR TITLE
[bugfix]: removing whitespace from malformed version ranges

### LINKED ISSUE #32: [Bug]: depRange.replaceAll is not a function
### Expected Behavior

I was expecting a failing report (this is a node.js v14.20.0 project with some really bad/messy dependencies)

### Actual Behavior

The following output:

```
$ npx depngn 18
Error: TypeError: depRange.replaceAll is not a function
    at /Users/redacted/.npm/_npx/80daaf8cfdca6619/node_modules/depngn/dist/cli.cjs:7609:27
    at step (/Users/redacted/.npm/_npx/80daaf8cfdca6619/node_modules/depngn/dist/cli.cjs:69:23)
    at Object.throw (/Users/redacted/.npm/_npx/80daaf8cfdca6619/node_modules/depngn/dist/cli.cjs:50:53)
    at rejected (/Users/redacted/.npm/_npx/80daaf8cfdca6619/node_modules/depngn/dist/cli.cjs:41:65)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
```

Looks like maybe one of the dependencies may be doing something weird with the `engines` field or something?

### Steps to Reproduce

1. Use node.js 14.x (yes, I know, it's well and truly out of date)
2. Run depngn: `npx depngn 18` on a project
3. The actual behaviour is observed

If you want a repository you can try this in, then see: https://github.com/inrupt/pod-browser

### Additional Information

I tried to add debug logging around the `depRange` line, and the only output it gave was `{ depRange: '>=6.9.0' }` before failing. I tried doing a `typeof` on `depRange` and it said "string", so I did a `typeof depRange.replaceAll` and it said `undefined`.

It looks like in node.js 14.x, `String.prototype.replaceAll` does not yet exist (it's a newer language feature): 

```
Welcome to Node.js v14.20.0.
Type ".help" for more information.
> String.prototype.replaceAll
undefined
>
```

So if you want to support Node.js 14.x then you'll need to add a polyfill in.

### What package manager are you seeing the problem with?

npm

### Relevant log output

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by kindoflew**
Resolves #32 

Wow, this was supposed to be a quick swap out of `replaceAll`, but I realized that it isn't actually that simple (when is it ever? 😅)

# Context
Initially, the `range.replaceAll(' ', '')` call was there to handle malformed version strings (like if there is a space between `>=` and the actual number). However, the comments in [this issue](https://github.com/omichelsen/compare-versions/issues/63) in `compare-versions` made me realize that this isn't actually a good idea.

TL;DR -- whitespace actually has semantic meaning in certain version ranges.

Take the following version range, for example:

```
1.2.7 || >=1.2.9 <2.0.0
```
The whitespace between `>=1.2.9` and `<2.0.0` represents a logical AND. if we stripped that out, it'd be a newly malformed version range. so, following @arielj's suggestion, i went with RegEx.

The RegEx removes whitespace in four situations:
1. Between the characters of the comparator operators (ie, `< =`)
2. Between the operators and the version number (ie, `^ 12`)
3. Between the digits and decimals in the version number (ie, `1 2 .0 . 0`)
4. Between two versions in a logical AND range (ie, `>=4.2.0     <=8.0.0`)

(in the last case, it leaves one space, so each version can be asserted separately)

I believe this is _good enough_ for now (thanks Sandi Metz 😂).

Not to mention, if you read through the above-linked issue in `compare-versions`, the author is currently working on supporting this out-of-the-box, so we can stop worrying about it when the new major version is released.

# My RegEx ~might~ probably does suck

I tested it on regex101.com and it works. but i'm sure it could be optimized somehow. it's long and ugly and has lots of lookaheads/lookbehinds and logical ORs.

Also, it doesn't take into account whitespace in the "extended portions" of a version (ie, `12.0.0-rc.1.canary.beta`) but that probably shouldn't come up for Node versions (I think?).

Another thought is maybe handling malformed version ranges more gracefully (instead of throwing an error). maybe we can have it output something like:
```
 {
  compatible: 'malformed',
  range: '1 2 . 0 . 0'
}
```

I figured that was outside the scope of this PR though. This should fix the issue for now, for the majority of cases (and then it can be `compare-versions`' problem 😂)

# Details

Instead of using if/else logic when there's a `||` present, we treat every range the same:
1. split at `||` (if present)
2. remove whitespace using that obnoxiously long RegEx
3. split the remaining range at `' '` (if present)
4. assert every remaining range with `satisfies`

# QA

To confirm the original bug is gone, just downgrade your Node version to something `v14` or below and run `dev`. It should pass now.

To confirm the RegEx works, I guess you could use an online regex tester for weird edge cases. Or set up a mock project with dependencies that have weird ranges specified.

### CODE REVIEW

File: src/queries/getPackageData.ts [hunk 1]
```diff
@@ -13,14 +13,26 @@ function isCompatible(nodeVersion: string, depRange: string) {
...
+// trims leading and trailing whitespace, whitespace
+// between the comparator operators and the actual version number, 
+// and whitespace between the numbers/wildcards/decimals in the actual
+// version number. ie, ' > = 1 2. 0 .0 ' becomes '>=12.0.0'
+//
+// it also ensures there's only one space in logical `AND` ranges
+// ie, '>=1.2.9 <2.0.0', because we want to split those later
+function removeWhitespace(range: string) {
+  const extraSpaceRegEx =
+    /((?<=(<|>))(\s+)(?=(=)))|(?<=(<|>|=|\^|~))(\s+)(?=\d)|((?<=(\d|\.|\*|x|X))(\s+)(?=(\d|\.|\*|x|X)))|(?<=\d)(\s+)(?=\s<|>)/g;
```
Comment (arielj) [inline src/queries/getPackageData.ts]: this looks scary xD, I'd like to give it a try into changing this to something less scary, can you share some of the malformed strings and the expected output?

it might be a good idea to add tests for this function with input/output examples

cause I can imagine there's some other solution (maybe combining regexp with strip, or smaller regexps to replace different parts one at a time)

Comment (arielj) [reply src/queries/getPackageData.ts]: hi lew :D

Comment (kindoflew) [reply src/queries/getPackageData.ts]: unfortunately, JavaScript doesn't have a native `strip` for strings. we could break down the RegEx into each conditional one though and then just chain `replace` calls?

```typescript
const comparatorWhitespace = /((?<=(<|>))(\s+)(?=(=)))/g;
const comparatorAndVersionWhiteSpace = /(?<=(<|>|=|\^|~))(\s+)(?=\d)/g;
const versionWhitespace = /((?<=(\d|\.|\*|x|X))(\s+)(?=(\d|\.|\*|x|X)))/g;
const betweenWhitespace = /(?<=\d)(\s+)(?=\s<|>)/g;

return range
  .trim()
  .replace(comparatorWhitespace, '')
  .replace(comparatorAndVersionWhiteSpace, '')
  .replace(versionWhitespace, '')
  .replace(betweenWhitespace, '');
```

also, here are some test cases for things that would cause `satisfies` to throw:

```
> =12.0.0
> 12.0.0
1 2. 0 .0
>= 12.0.0
> = 1 2 . 0 . 0
> = 4 . 2 . 0     < 8. 0 .0
```

Comment (kindoflew) [reply src/queries/getPackageData.ts]: oohhh, i just came up with a waaay shorter RegEx using _negative_ lookahead:

```
/(\s)(?!(<|>|^|~))/g
```
That matches any whitespace that is not immediately followed by `< | > | ~ | ^`

It's not perfect (because I'm not sure if you can do an AND range like `>=4.2.0 8.0.0`, in which case this would remove the whitespace between them), but it should cover all the cases in the above comment.

Comment (KostiantynPopovych) [reply src/queries/getPackageData.ts]: Hey @kindoflew, I think we need to move forward with a fully working solution, as this `/(\s)(?!(<|>|^|~))/g` is not quite yet, let's consider the original one you suggested, for now. We can improve it in the future, but it will be good to have it merged asap.

Comment (kindoflew) [reply src/queries/getPackageData.ts]: @KostiantynPopovych, i'm _pretty_ sure that both RegExes will do the same thing -- neither one would cover something like `>=4.2.0 8.0.0`.

basically, we'd have to write another RegEx for SemVer ranges (which [isn't exactly straightforward](). and those are just for version numbers -- they don't include comparators) and then strip all but one whitespace between them.

i think i'm sort of leaning towards the shorter one, since it'll (hopefully) be a temporary fix until `compare-versions` releases the next major version (or `npm`'s `semver` package figures out how to get rid of it's cyclical dependency and we could just use that one instead -- it has all this stuff out of the box).

Comment (KostiantynPopovych) [reply src/queries/getPackageData.ts]: @kindoflew Oh, I see, I was playing with these solutions yesterday and probably messed something up because the original (long) regex was working for this `>=4.2.0 8.0.0` case, maybe I have fixed one thing but crashed another. So anyway, if we decide to move with any of them, we need to inform users about this issue, probably creating a section in `README.md` named `Known Issues` and describing this behavior there.

How does it sound to you?

Comment (kindoflew) [reply src/queries/getPackageData.ts]: @KostiantynPopovych i figured out one last thing to make `'>=4.2.0     8.0.0'` work.

in `safeSatisfies` after we `split(' ')`, we filter out any extra whitespace between them (`filter((r) => !!r)`) and that should leave us with two individual ranges, no matter what. the whole thing looks like:

```javascript
range
  .split(' ')
  .filter((r) => !!r)
  .every((r) => satisfies(nodeVersion, r))
```

i also did a very small amount of testing execution speed between long regex, split long regex, and the short regex (i surrounded the `forEach` loop that calls `getPackageData` with `console.time` calls, tested with the dependencies of `depngn`). all 3 were pretty consistently between 1ms and 2ms (with long regex the only one to go _below_ 1ms and split long regex the only one to go _above_ 2ms). tests weren't very rigorous, but they seem to take about the same amount of time, long regex having a slight edge.

i'm down to do whatever you'd like! i'm leaning towards long regex, with maybe a comment about what each chunk represents? either way, that `filter` i mentioned solves the last whitespace problem, so 🎉

Comment (KostiantynPopovych) [reply src/queries/getPackageData.ts]: @kindoflew, sorry for the long reply! In the case of adding an additional filter - `range.trim()` sounds redundant, coz we will filter these white spaces in the following function, isn't it?

Comment (kindoflew) [reply src/queries/getPackageData.ts]: @KostiantynPopovych  oh yeah, you're right. i thought i solved something else but i didn't 😅. i have too many versions of this locally.

anyway, the issue isn't with `>=4.2.0 8.0.0`, it's with `>=4.2.0      8.0.0`.

but, i think i finally figured it out (i didn't benchmark execution time, but with the optimizations in #39, i don't think it'll matter much. plus i've been looking at how big libraries like `axios` and `semver` do regex stuff, and they seem to have multiple `replace` and `match` calls all over the place and do just fine).

anyways, this (hopefully) is the final version i came up with:

```typescript
function removeWhitespace(range: string) {
  const comparatorWhitespace = /((?<=(<|>))(\s+)(?=(=)))/g;
  const comparatorAndVersionWhiteSpace = /(?<=(<|>|=|\^|~))(\s+)(?=\d)/g;
  const remainingRange =
   /(((\d\s*){1,3})\.\s*(\d\s*){1,3}\.\s*((0\s*)|(\d\s*){1,2})(?!(((<|>)=?)|~|^)?(((\d\s*){1,3})\.)|(\d)))/g;
  return range
    .trim()
    .replace(comparatorWhitespace, '')
    .replace(comparatorAndVersionWhiteSpace, '')
    .replace(remainingWhitespace, (match: string) => match.replace(/\s+/g, ''));
}
```

The first two regexes (`comparatorWhitespace` and `comparatorAndVersionWhitespace`) strip out the whitespace between what the names say. this would leave us with things like:

```
>=4.2.0      >=18.0.0
>=4.2. 0     1 8 .0.0   
>=1 2.  0.0
>1 2.0.1 0    1
1 2 . 1 0 . 0 
>=1 2.0.1 0
>=1   2.0.0
>=4.2.0      8.0
>=4.2.1  0      8
```

The last regex (`remainingRange`) matches a single SemVer version that may or may not have whitespace between digits or digits and decimals. and then we replace it with the same string with whitespace removed. this leaves us with either a whitespace-less single range or an AND range with a single space in between. we still need the last `filter` call because the last regex will not match on something like `>1    <3`, which is a valid range (minus the extra whitespace).

Comment (kindoflew) [reply src/queries/getPackageData.ts]: @KostiantynPopovych however, looking at what `node-semver` does, they assume the only additional spaces will be between comparator operator characters or between comparator operators and the version number -- they consider additional space between digits/decimals to be malformed and just throw an error.

we could do that (but instead of throwing, maybe we just return `{ compatible: undefined, range: {malformedRange} }` or maybe even `{ compatible: 'invalid', range: {malformedRange} }`. what do you think of that?

Comment (KostiantynPopovych) [reply src/queries/getPackageData.ts]: @kindoflew this `{ compatible: 'invalid', range: {malformedRange} }` looks great to me! I think it will be good to ask why `node-semver` decide to mark it as malformed 🤔

Comment (kindoflew) [reply src/queries/getPackageData.ts]: i think because the RegEx to parse it is impossible? or at least computationally expensive.

consider the case where you have the following, malformed range:

```
<1   2
```

is that supposed to be "less than 12" or "less than 1 AND exactly 2"? i think because there are so many different variations for SemVer strings, you kind of have to follow the happy path for the most part and just throw when you can't be sure. at least that's my assumption from researching this stuff the past few weeks ¯\\\_(ツ)_/¯

Comment (KostiantynPopovych) [reply src/queries/getPackageData.ts]: Sounds reasonable) So, I believe we can go with this approach.

**Comment by KostiantynPopovych** [comment]
Great job, @kindoflew, and thank you!
